processing/src/main/java/io/druid/query/join/DataSourceJoinInputSpec.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.join;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.collect.Iterables;
+import io.druid.query.DataSource;
+import io.druid.query.Queries;
+import io.druid.query.spec.QuerySegmentSpec;
+import org.joda.time.Duration;
+
+import java.util.Objects;
+
+public class DataSourceJoinInputSpec implements JoinInputSpec
+{
+  private final DataSource dataSource;
+  private final QuerySegmentSpec querySegmentSpec;
+  private volatile Duration duration;
```

Comment:
It's safe to make it non-volatile because Joda-time ensures "final" semantics for it's basic immutable classes: http://cs.oswego.edu/pipermail/concurrency-interest/2011-June/007976.html